File: LeetCode Problems/1 to 100/17. Letter Combinations of a Phone Number/17. Letter Combinations of a Phone Number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if (digits.empty())
            return {};
        vector<string> str{"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        vector<string> comb;
        if (digits.size() == 1) {
            string s = str[digits[0] - '0' - 2];
            for (int i = 0; i < s.size(); i++) {
                string word;
                word.push_back(s[i]);
                comb.push_back(word);
            }
        } else if (digits.size() == 2) {
            string s1 = str[digits[0] - '0' - 2];
            string s2 = str[digits[1] - '0' - 2];
            for (int i = 0; i < s1.size(); i++) {
                for (int j = 0; j < s2.size(); j++) {
                    string word;
                    word.push_back(s1[i]);
                    word.push_back(s2[j]);
                    comb.push_back(word);
                }
            }
        } else if (digits.size() == 3) {
            string s1 = str[digits[0] - '0' - 2];
            string s2 = str[digits[1] - '0' - 2];
            string s3 = str[digits[2] - '0' - 2];
            for (int i = 0; i < s1.size(); i++) {
                for (int j = 0; j < s2.size(); j++) {
                    for (int k = 0; k < s3.size(); k++) {
                        string word;
                        word.push_back(s1[i]);
                        word.push_back(s2[j]);
                        word.push_back(s3[k]);
                        comb.push_back(word);
                    }
                }
            }
        } else {
            string s1 = str[digits[0] - '0' - 2];
            string s2 = str[digits[1] - '0' - 2];
            string s3 = str[digits[2] - '0' - 2];
            string s4 = str[digits[3] - '0' - 2];
            for (int i = 0; i < s1.size(); i++) {
                for (int j = 0; j < s2.size(); j++) {
                    for (int k = 0; k < s3.size(); k++) {
                        for (int l = 0; l < s4.size(); l++) {
                            string word;
                            word.push_back(s1[i]);
                            word.push_back(s2[j]);
                            word.push_back(s3[k]);
                            word.push_back(s4[l]);
                            comb.push_back(word);
                        }
                    }
                }
            }
        }
        return comb;
    }
};

File: LeetCode Problems/1 to 100/17. Letter Combinations of a Phone Number/17. Letter Combinations of a Phone Number.java
class Solution {
    public List<String> letterCombinations(String digits) {
        int n = digits.length();
        if (n == 0) return new ArrayList<String>();
        List<String> list = new ArrayList<>(List.of("abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"));
        List<String> ans = new ArrayList<>();
        if (n == 1) {
            String s = list.get((digits.charAt(0) - '0') - 2);
            for (int i = 0; i < s.length(); i++) {
                ans.add((s.charAt(i) + ""));
            }
        } else if (n == 2) {
            String s1 = list.get((digits.charAt(0) - '0') - 2);
            String s2 = list.get((digits.charAt(1) - '0') - 2);
            for (int i = 0; i < s1.length(); i++) {
                for (int j = 0; j < s2.length(); j++) {
                    String word = "";
                    word += s1.charAt(i);
                    word += s2.charAt(j);
                    ans.add(word);
                }
            }
        } else if (n == 3) {
            String s1 = list.get((digits.charAt(0) - '0') - 2);
            String s2 = list.get((digits.charAt(1) - '0') - 2);
            String s3 = list.get((digits.charAt(2) - '0') - 2);
            for (int i = 0; i < s1.length(); i++) {
                for (int j = 0; j < s2.length(); j++) {
                    for (int k = 0; k < s3.length(); k++) {
                        String word = "";
                        word += s1.charAt(i);
                        word += s2.charAt(j);
                        word += s3.charAt(k);
                        ans.add(word);
                    }
                }
            }
        } else if (n == 4) {
            String s1 = list.get((digits.charAt(0) - '0') - 2);
            String s2 = list.get((digits.charAt(1) - '0') - 2);
            String s3 = list.get((digits.charAt(2) - '0') - 2);
            String s4 = list.get((digits.charAt(3) - '0') - 2);
            for (int i = 0; i < s1.length(); i++) {
                for (int j = 0; j < s2.length(); j++) {
                    for (int k = 0; k < s3.length(); k++) {
                        for (int l = 0; l < s4.length(); l++) {
                            String word = "";
                            word += s1.charAt(i);
                            word += s2.charAt(j);
                            word += s3.charAt(k);
                            word += s4.charAt(l);
                            ans.add(word);
                        }
                    }
                }
            }
        }
        return ans;
    }
}
